### PR TITLE
Implement DNS forwarding API

### PR DESCRIPTION
**Pull request by Miciah**
#### Bump openshift/api for DNS forwarding API

Bump to github.com/openshift/api@610e47be70dbf6a15ca0818d09f513f5aed60753 to get the DNS forwarding API.

* `go.mod`: Bump.
* `go.sum`:
* `manifests/00-custom-resource-definition.yaml`:
* `vendor/github.com/openshift/api/*`:
* `vendor/modules.txt`: Regenerate.

#### Implement DNS forwarding API

* `pkg/manifests/manifests.go` (`DNSConfigMapAsset`): Delete constant.
(`DNSConfigMap`): Delete function.
* `pkg/manifests/manifests_test.go` (`TestManifests`): Delete `DNSConfigMap`.
* `pkg/operator/controller/controller_dns_configmap.go` (`corefileTemplate`): New variable containing a Corefile template.
(`corefileTemplateFunctions`): New template function map.
(`translatePolicyName`): New function.  Translate DNS policy from the API name to the Corefile string.  Used in `corefileTemplateFunctions`.
(`ensureDNSConfigMap`): Handle errors from `desiredDNSConfigMap`.
(`desiredDNSConfigMap`): Add error return value.  Replace `DNSConfigMap` from the manifests package by using `corefileTemplate` and `corefileTemplateFunctions` to generate a Corefile from the parameters in the given DNS.
* `pkg/operator/controller/controller_dns_configmap_test.go:` New file.
(`TestDesiredDNSConfigmap`): New test.

### CODE REVIEW

**Comment by danehans** [comment]
Requires https://github.com/openshift/api/pull/441

File: go.sum [hunk 1]
```diff
@@ -221,6 +221,8 @@ github.com/openshift/api v3.9.1-0.20191002145753-bb291c9def6c+incompatible h1:HF
 github.com/openshift/api v3.9.1-0.20191002145753-bb291c9def6c+incompatible/go.mod h1:dh9o4Fs58gpFXGSYfnVxGR9PnV53I8TW84pQaJDdGiY=
 github.com/openshift/api v3.9.1-0.20191028134408-abf28f93703b+incompatible h1:1mW1j1mik2++Z+ehEaqm2xcWGiGYJUK8SdbkTgTS+N4=
 github.com/openshift/api v3.9.1-0.20191028134408-abf28f93703b+incompatible/go.mod h1:dh9o4Fs58gpFXGSYfnVxGR9PnV53I8TW84pQaJDdGiY=
+github.com/openshift/api v3.9.1-0.20191105205100-610e47be70db+incompatible h1:HVPbu7fa4U8JUNuLNkXOXkhl/YvVk9xgj1r0DRy902Y=
+github.com/openshift/api v3.9.1-0.20191105205100-610e47be70db+incompatible/go.mod h1:dh9o4Fs58gpFXGSYfnVxGR9PnV53I8TW84pQaJDdGiY=
```
Comment (danehans) [inline go.sum]: @Miciah  do you think the `
[sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] works for CRD with validation schema [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s] ` failure is associated to the api bump?

Why are two `github.com/openshift/api` required here?

Comment (Miciah) [reply go.sum]: > @Miciah do you think the `[sig-api-machinery] CustomResourcePublishOpenAPI [Privileged:ClusterAdmin] works for CRD with validation schema [Conformance] [Suite:openshift/conformance/parallel/minimal] [Suite:k8s]` failure is associated to the api bump?

No, I see several reports for the same failure:

https://bugzilla.redhat.com/show_bug.cgi?id=1760198
https://bugzilla.redhat.com/show_bug.cgi?id=1754091
https://bugzilla.redhat.com/show_bug.cgi?id=1761043

Furthermore, https://search.svc.ci.openshift.org/?search=works%20for%20CRD%20with%20validation%20schema shows 161 matches.

> Why are two `github.com/openshift/api` required here?

`go mod` added them.  I'll see whether I can remove them...

Comment (Miciah) [reply go.sum]: I ran `go mod tidy`, which cleaned out a lot of old entries from `go.mod`.

Comment (Miciah) [reply go.sum]: As for why there are two lines for the same package, https://golang.org/cmd/go/#hdr-Module_authentication_using_go_sum explains:

> The first line gives the hash of the module version's file tree. The second line appends "/go.mod" to the version and gives the hash of only the module version's (possibly synthesized) go.mod file.

**Pull request merged**